File: android/app/src/main/java/com/speakpro/features/review/NotificationsViewModel.kt
package com.speakpro.features.review

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.speakpro.core.network.ApiService
import com.speakpro.data.models.NotificationItem
import com.speakpro.data.models.NotificationPrefs
import com.speakpro.data.models.UpdatePrefsRequest
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltViewModel
class NotificationsViewModel @Inject constructor(
    private val apiService: ApiService,
) : ViewModel() {

    private val _items = MutableStateFlow<List<NotificationItem>>(emptyList())
    val items: StateFlow<List<NotificationItem>> = _items.asStateFlow()

    private val _unread = MutableStateFlow(0)
    val unread: StateFlow<Int> = _unread.asStateFlow()

    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()

    private val _errorMessage = MutableStateFlow<String?>(null)
    val errorMessage: StateFlow<String?> = _errorMessage.asStateFlow()

    fun load() {
        viewModelScope.launch {
            _isLoading.value = true
            _errorMessage.value = null
            try {
                val resp = apiService.getNotifications()
                _items.value = resp.data?.items ?: emptyList()
                _unread.value = resp.data?.unread ?: 0
            } catch (e: Exception) {
                _errorMessage.value = e.localizedMessage
            } finally {
                _isLoading.value = false
            }
        }
    }

    fun markAllRead() {
        viewModelScope.launch {
            try {
                apiService.markAllNotificationsRead()
                _items.value = _items.value.map { it.copy(isRead = true) }
                _unread.value = 0
            } catch (e: Exception) {
                _errorMessage.value = e.localizedMessage
            }
        }
    }

    fun markRead(id: String) {
        val idx = _items.value.indexOfFirst { it.id == id }
        if (idx < 0 || _items.value[idx].isRead) return
        viewModelScope.launch {
            try {
                apiService.markNotificationRead(id)
                _items.value = _items.value.mapIndexed { i, item ->
                    if (i == idx) item.copy(isRead = true) else item
                }
                _unread.value = (_unread.value - 1).coerceAtLeast(0)
            } catch (e: Exception) {
                _errorMessage.value = e.localizedMessage
            }
        }
    }
}

@HiltViewModel
class NotificationPrefsViewModel @Inject constructor(
    private val apiService: ApiService,
) : ViewModel() {

    private val _prefs = MutableStateFlow(
        NotificationPrefs(userId = "", quietStart = "22:30:00", quietEnd = "07:30:00", pushEnabled = true),
    )
    val prefs: StateFlow<NotificationPrefs> = _prefs.asStateFlow()

    private val _isSaving = MutableStateFlow(false)
    val isSaving: StateFlow<Boolean> = _isSaving.asStateFlow()

    private val _errorMessage = MutableStateFlow<String?>(null)
    val errorMessage: StateFlow<String?> = _errorMessage.asStateFlow()

    fun load() {
        viewModelScope.launch {
            try {
                val resp = apiService.getNotificationPrefs()
                val data = resp.data
                if (data != null) _prefs.value = data
            } catch (_: Exception) {
                // 首次没记录，用默认值
            }
        }
    }

    fun updateQuietStart(s: String) { _prefs.value = _prefs.value.copy(quietStart = s) }
    fun updateQuietEnd(s: String)   { _prefs.value = _prefs.value.copy(quietEnd = s) }
    fun updatePushEnabled(b: Boolean) { _prefs.value = _prefs.value.copy(pushEnabled = b) }

    fun save(onDone: () -> Unit) {
        viewModelScope.launch {
            _isSaving.value = true
            _errorMessage.value = null
            try {
                apiService.updateNotificationPrefs(
                    UpdatePrefsRequest(
                        quietStart = _prefs.value.quietStart,
                        quietEnd = _prefs.value.quietEnd,
                        pushEnabled = _prefs.value.pushEnabled,
                    ),
                )
                onDone()
            } catch (e: Exception) {
                _errorMessage.value = e.localizedMessage
            } finally {
                _isSaving.value = false
            }
        }
    }
}

File: ios/SpeakPro/Features/Review/NotificationsViewModel.swift
import Foundation

@MainActor
final class NotificationsViewModel: ObservableObject {

    @Published var items: [NotificationItem] = []
    @Published var unread: Int = 0
    @Published var isLoading: Bool = false
    @Published var errorMessage: String? = nil

    func load() async {
        isLoading = true; defer { isLoading = false }
        do {
            let resp = try await APIClient.shared.getNotifications()
            items = resp.items
            unread = resp.unread
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func markAllRead() async {
        do {
            _ = try await APIClient.shared.markAllNotificationsRead()
            // 本地直接把 isRead 置为 true（服务端已改，避免再拉一次）
            items = items.map { Self.markedRead($0) }
            unread = 0
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func markRead(id: String) async {
        // 已读则跳过
        guard let i = items.firstIndex(where: { $0.id == id }), !items[i].isRead else { return }
        do {
            _ = try await APIClient.shared.markNotificationRead(id: id)
            items[i] = Self.markedRead(items[i])
            unread = max(0, unread - 1)
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    private static func markedRead(_ item: NotificationItem) -> NotificationItem {
        NotificationItem(
            id: item.id,
            userId: item.userId,
            kind: item.kind,
            title: item.title,
            body: item.body,
            payload: item.payload,
            isRead: true,
            createdAt: item.createdAt,
        )
    }
}

@MainActor
final class NotificationPrefsViewModel: ObservableObject {

    @Published var quietStart: String = "22:30:00"
    @Published var quietEnd: String = "07:30:00"
    @Published var pushEnabled: Bool = true
    @Published var isSaving: Bool = false
    @Published var errorMessage: String? = nil

    func load() async {
        do {
            let p = try await APIClient.shared.getNotificationPrefs()
            quietStart = p.quietStart
            quietEnd = p.quietEnd
            pushEnabled = p.pushEnabled
        } catch {
            // 首次无记录 —— 用默认值即可
        }
    }

    func save() async {
        isSaving = true; defer { isSaving = false }
        errorMessage = nil
        do {
            _ = try await APIClient.shared.updateNotificationPrefs(
                UpdatePrefsRequest(
                    quietStart: quietStart,
                    quietEnd: quietEnd,
                    pushEnabled: pushEnabled,
                ),
            )
        } catch {
            errorMessage = error.localizedDescription
        }
    }
}
